server/configuration.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+// configuration captures the plugin's external configuration as exposed in the Mattermost server
+// configuration, as well as values computed from the configuration. Any public fields will be
+// deserialized from the Mattermost server configuration in OnConfigurationChange.
+//
+// As plugins are inherently concurrent (hooks being called asynchronously), and the plugin
+// configuration can change at any time, access to the configuration must be synchronized. The
+// strategy used in this plugin is to guard a pointer to the configuration, and clone the entire
+// struct whenever it changes. You may replace this with whatever strategy you choose.
+type configuration struct {
+	// AWS access key
+	AWSAccessKeyID string
+
+	// AWS secret key
+	AWSSecretAccessKey string
+
+	// AWS region with "us-east-1" as default
+	AWSRegion string
+
+	// disable plugin
+	disabled bool
+}
+
+// Clone deep copies the configuration. Your implementation may only require a shallow copy if
+// your configuration has no reference types.
+func (c *configuration) Clone() *configuration {
+	return &configuration{
+		AWSAccessKeyID:     c.AWSAccessKeyID,
+		AWSSecretAccessKey: c.AWSSecretAccessKey,
+		AWSRegion:          c.AWSRegion,
+		disabled:           c.disabled,
+	}
+}
+
+// getConfiguration retrieves the active configuration under lock, making it safe to use
+// concurrently. The active configuration may change underneath the client of this method, but
+// the struct returned by this API call is considered immutable.
+func (p *Plugin) getConfiguration() *configuration {
+	p.configurationLock.RLock()
+	defer p.configurationLock.RUnlock()
+
+	if p.configuration == nil {
+		return &configuration{}
+	}
+
+	return p.configuration
+}
+
+// setConfiguration replaces the active configuration under lock.
+//
+// Do not call setConfiguration while holding the configurationLock, as sync.Mutex is not
+// reentrant. In particular, avoid using the plugin API entirely, as this may in turn trigger a
+// hook back into the plugin. If that hook attempts to acquire this lock, a deadlock may occur.
+//
+// This method panics if setConfiguration is called with the existing configuration. This almost
+// certainly means that the configuration was modified without being cloned and may result in
+// an unsafe access.
+func (p *Plugin) setConfiguration(configuration *configuration) {
+	p.configurationLock.Lock()
+	defer p.configurationLock.Unlock()
+
+	if configuration != nil && p.configuration == configuration {
+		panic("setConfiguration called with the existing configuration")
+	}
+
+	p.configuration = configuration
+}
+
+func (p *Plugin) diffConfiguration(newConfiguration *configuration) {
```

Comment:
Is this method needed?

Reply:
I just removed it now but let me know what's the proper approach.

Reply:
Is an unneeded method. You can just drop it